plugins.ustreamtv: fix issue with muxed hls streams

beardypig opened the pull request:
Should fix #1767... changes how the HLS stream is wrapped so that it should support any future HLS stream changes.

back-to commented:
> http://www.ustream.tv/channel/18525677 

it starts but 500 error

```
[stream.hls][error] Failed to open segment 11: Unable to open URL: ...
(500 Server Error: Internal Server Error for url: ... )
```

---

> http://www.ustream.tv/channel/live-iss-stream

the plugin itself is very unstable,
it works only once every ~20 tries

beardypig commented:
That channel did generate a lot of errors, I personally don’t have many problems - but I only really use a couple of streams. I will try to do some debugging though.

back-to commented:
the first stream is probably not fixable with hls,
so it won't work with streamlink

---

`live-iss-stream`
might be an ustream issue aswell
or at least it's a different issue than this PR

liamkennedy commented:
Does this issue explain why I can only play out about 15 seconds of a Ustream channel before it freezes?  Youtube live streams seem to be ok.  

Testing this on 0.13.0 (and tried it on some older versions too just to see)

These are example streams to see the issue - Ustream and Youtube

**TWiT / Leo Laporte**
Ustream (**freezes**)
`streamlink http://ustream.tv/leolaporte best --player omxplayer --fifo -l debug`
Youtube (works)
`streamlink https://www.youtube.com/watch?v=Zs5Z2EGqJFw best --player omxplayer --fifo -l debug

Or NASA's HDEV stream (will be blank regularly - and no audio so a little more difficult to use to help figure this out).  
**Ustream (freezes)**
`streamlink http://www.ustream.tv/channel/iss-hdev-payload best --player "omxplayer" --fifo -l debug`
**Youtube (works)**  - By the way... yes.. NASA is now streaming this via Youtube - just added last week.  
`streamlink https://www.youtube.com/watch?v=625ZC_KpweI best --player omxplayer --fifo -l debug`

The debug output is as expected.  For Ustream you see a few successful segment downloads initially followed by Reloading playlist messages.. then it eventually quits.  For Youtube.. nothing unexpected.. it just keeps loading the segments as expected.  

Thanks for working on this.

beardypig commented:
Hi @liamkennedy, could you try with the extra option `--hls-live-edge 10`, I have had success using that option. The default is `--hls-live-edge 3` and that doesn't seem to work with the UStream streams... It will mean that you're about 60 seconds behind live though - but I think this is something you might have to live with. A lower value might also work, but I didn't experiment enough.

liamkennedy commented:
@beardypig Well.. that helps... but ultimately does not change the result.  The live video seems to buffer a lot more segments right at the start - but then it still goes back in to the "reloading playlist" and then quits.  It extends the video from 15-20 seconds to about 30-40 seconds play time.  

Thanks for the suggestion though.

liamkennedy commented:
@beardypig - yes - I can test with the new changes.  Are these changes just in the ustreamtv.py file as posted with this PR or should I pull down the full current development version.  (hint: I'm not all that savvy with github... sorry!)

beardypig commented:
Yep, just that file :)

liamkennedy commented:
I modifed the debug statements to prefix some with "bp" (just so I could be certain I was running with your changes)... and this is the output (the Reloading playlist goes on for a bit more finishing with the usual final timeout)

```
[console][debug] OS:         Linux-4.14.34-v7+-armv7l-with-debian-9.4
[console][debug] Python:     2.7.13
[console][debug] Streamlink: 0.13.0
[console][debug] Requests(2.18.4), Socks(1.6.7), Websocket(0.48.0)
[console][info] Found matching plugin ustreamtv for URL http://ustream.tv/leolaporte
[plugin.ustreamtv][debug] Searching for media ID on the page
[plugin.ustreamtv][debug] Connecting to UStream API: media_id=1524, application=channel, referrer=http://ustream.tv/leolaporte, cluster=live, app_id=11, app_ver=2
[plugin.ustreamtv][debug] bp Got new host=http://c742ec95-lp-omega.ums.ustream.tv, and connectionId=p5g0ap3v7a3raijkmr2for7a1d-10
[plugin.ustreamtv][debug] bp poll took 0.17s
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[console][info] Available streams: 252p (worst), 360p, 486p, 720p (best)
[console][info] Opening stream: 720p (ustream)
[plugin.ustreamtv][debug] Starting API polling thread
[stream.hls][debug] Reloading playlist
[stream.hls][debug] First Sequence: 1301; Last Sequence: 1309
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 1301; End Sequence: None
[stream.hls][debug] Adding segment 1301 to queue
[console][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 1302 to queue
[stream.hls][debug] Adding segment 1303 to queue
[stream.hls][debug] Adding segment 1304 to queue
[stream.hls][debug] Adding segment 1305 to queue
[stream.hls][debug] Adding segment 1306 to queue
[stream.hls][debug] Adding segment 1307 to queue
[stream.hls][debug] Adding segment 1308 to queue
[stream.hls][debug] Adding segment 1309 to queue
[stream.hls][debug] Download of segment 1301 complete
[console][info] Creating pipe streamlinkpipe-1551
[console][info] Starting player: omxplayer
[stream.hls][debug] Download of segment 1302 complete
[console][debug] Writing stream to output
[stream.hls][debug] Download of segment 1303 complete
[plugin.ustreamtv][debug] bp poll took 0.04s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Download of segment 1304 complete
[stream.hls][debug] Download of segment 1305 complete
[stream.hls][debug] Download of segment 1306 complete
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Download of segment 1307 complete
[stream.hls][debug] Download of segment 1308 complete
[stream.hls][debug] Download of segment 1309 complete
[plugin.ustreamtv][debug] bp poll took 0.04s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.12s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.82s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.07s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] bp poll took 0.04s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] bp poll took 0.04s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.05s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll took 0.04s
[plugin.ustreamtv][debug] bp poll response: cmd
[plugin.ustreamtv][debug] bp poll response: cmd

```

liamkennedy commented:
Although... now testing on NASA's HDEV channel - your fix is working!!!

So.. that's actually what I really want to have working and it's great that your change fixes that.. but then.. why doesn't the Leolaporte channel work too.  Very strange.

liamkennedy commented:
Although #2 - now NASA HDEV is back to doing what the Leolaporte channel was doing too.. so it's still freezing.  Iincreasing the --hls-live-edge to 20 still get the same reloading message and timeout eventually

So... getting inconsistent results.  darn.

beardypig commented:
@liamkennedy just to be sure, can you replace this whole file with this one https://raw.githubusercontent.com/beardypig/streamlink/c82c3395efbd08f119941ffff98892fffa66e096/src/streamlink/plugins/ustreamtv.py ? Some of the debug messages don’t look as they should, and one is missing in your log.

liamkennedy commented:
Just replaced it with that one... and to confirm I did a diff and could see the  version I had did not include some of the changes in this one.  Alas.. I still get the same result.  Logs below up to the point when the Reloading playlist message comes up with no more adding segment messages.  

```
pi@issabove:~/issabove/tmp $ streamlink http://www.ustream.tv/channel/iss-hdev-payload best --player omxplayer --fifo -l debug --hls-live-edge 10
[console][debug] OS:         Linux-4.14.34-v7+-armv7l-with-debian-9.4
[console][debug] Python:     2.7.13
[console][debug] Streamlink: 0.13.0
[console][debug] Requests(2.18.4), Socks(1.6.7), Websocket(0.48.0)
[console][info] Found matching plugin ustreamtv for URL http://www.ustream.tv/channel/iss-hdev-payload
[plugin.ustreamtv][debug] Searching for media ID on the page
[plugin.ustreamtv][debug] Connecting to UStream API: media_id=17074538, application=channel, referrer=http://www.ustream.tv/channel/iss-hdev-payload, cluster=live, app_id=11, app_ver=2
[plugin.ustreamtv][debug] BP2 Got new host=http://c742ec95-lp-omega.ums.ustream.tv, and connectionId=28lpt5stn2dphk8l5ojjdpp383-16
[plugin.ustreamtv][debug] poll took 0.15s
[plugin.ustreamtv][debug] stream: {u'url': u'http://c742ec95-lp-omega.ums.ustream.tv/playlist/auditorium/channel/17074538/playlist.m3u8?token=6168863_1528846990285&appType=11&appVersion=2&ts=1528846990&chunkingType=improved&geo=US&sgn=f59aa0b5f4ba5ae9e7313063c6de51832cb20a79&preferredBitrate=0&cdnHost=uhs-akamai.ustream.tv/', u'name': u'uhls'}
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[console][info] Available streams: 252p (worst), 360p, 486p, 720p (best)
[console][info] Opening stream: 720p (ustream)
[plugin.ustreamtv][debug] Starting API polling thread
[stream.hls][debug] Reloading playlist
[stream.hls][debug] First Sequence: 28221; Last Sequence: 28230
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 28221; End Sequence: None
[stream.hls][debug] Adding segment 28221 to queue
[console][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 28222 to queue
[stream.hls][debug] Adding segment 28223 to queue
[stream.hls][debug] Adding segment 28224 to queue
[stream.hls][debug] Adding segment 28225 to queue
[stream.hls][debug] Adding segment 28226 to queue
[stream.hls][debug] Adding segment 28227 to queue
[stream.hls][debug] Adding segment 28228 to queue
[stream.hls][debug] Adding segment 28229 to queue
[stream.hls][debug] Adding segment 28230 to queue
[console][info] Creating pipe streamlinkpipe-1072
[console][info] Starting player: omxplayer
[stream.hls][debug] Download of segment 28221 complete
[stream.hls][debug] Download of segment 28222 complete
[plugin.ustreamtv][debug] poll took 0.05s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848564, u'allViewers': 280205}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 669}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848567, u'allViewers': 280208}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 666}]}
[console][debug] Writing stream to output
[stream.hls][debug] Download of segment 28223 complete
[stream.hls][debug] Download of segment 28224 complete
[plugin.ustreamtv][debug] poll took 0.04s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848568, u'allViewers': 280209}]}
[stream.hls][debug] Download of segment 28225 complete
[stream.hls][debug] Download of segment 28226 complete
[stream.hls][debug] Download of segment 28227 complete
[plugin.ustreamtv][debug] poll took 0.04s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 667}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848570, u'allViewers': 280211}]}
[stream.hls][debug] Download of segment 28228 complete
[plugin.ustreamtv][debug] poll took 0.07s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 669}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848573, u'allViewers': 280214}]}
[stream.hls][debug] Download of segment 28229 complete
[stream.hls][debug] Download of segment 28230 complete
[plugin.ustreamtv][debug] poll took 0.05s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848575, u'allViewers': 280216, u'viewers': 672}]}
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] poll took 0.05s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848577, u'allViewers': 280218}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 674}]}
[plugin.ustreamtv][debug] poll took 0.04s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 678}]}
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848581, u'allViewers': 280222}]}
[plugin.ustreamtv][debug] poll took 0.05s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848584, u'allViewers': 280225}]}
[plugin.ustreamtv][debug] poll took 0.05s
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 248848589, u'allViewers': 280230, u'viewers': 681}]}

```

beardypig commented:
That's annoying @liamkennedy... I cannot replicate the issue when using `--hls-live-edge 10` :( If you can try it from a different IP, or try adding `--http-header "User-Agent=Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0"` ... I'm running out of ideas :)

liamkennedy commented:
You mean you want me to try this via on some other network?  Like if I were to connect one of my Raspberry Pi through a hotspot so I am using a different ISP?  I can definitely do that.  I did try that BEFORE I reported the issue (my first comment above).  However that was before I implemented the changes in this PR.. so it would be a good test case now. 

Curious though.. if you think it's IP related are you saying the somehow Ustream is doing "some weird stuff".. perhaps rate limiting my IP or something?

I'll also try adding the http-header too.  I'll try anything. 

I'm very grateful for the ideas.  I'll get back to you as soon as I've done all you suggested.  

Thanks!

Billy2011 commented:
@beardypig I think the issue #1767 still occurs in all videos: [http://www.ustream.tv/recorded/36348141](http://www.ustream.tv/recorded/36348141)

Billy2011 commented:
@beardypig the VOD playback works only with the `-o` or the `--player-passthrough` option, if I try to start the stream with the std playback option, then the VLC player opens, but the stream doesn't start and the player will close after a moment:
```
streamlink http://www.ustream.tv/recorded/41382414 best -l debug
[console][debug] OS:         Windows 8
[console][debug] Python:     2.7.9
[console][debug] Streamlink: 0.13.0
[console][debug] Requests(2.18.4), Socks(1.6.7), Websocket(0.47.0)
[console][info] Found matching plugin ustreamtv for URL http://www.ustream.tv/recorded/41382414
[plugin.ustreamtv][debug] Connecting to UStream API: media_id=41382414, application=recorded, referrer=http://www.ustream.tv/recorded/41382414, cluster=live, app_id=11, app_ver=2
[plugin.ustreamtv][debug] Got new host=http://12901cc3-lp-live.ums.ustream.tv, and connectionId=515lselsb7l3aom30uvsmtfbgm-5
[plugin.ustreamtv][debug] poll took 0.85s
[plugin.ustreamtv][debug] stream: {u'streams': [{u'mimetype': u'video/mp4', u'streamName': u'http://tcdn.ustream.tv/video/41382414?preset_id=2&e=1528903019&h=606146e5a847f343ff0d60d78d803678'}], u'name': u'ustream'}
[console][info] Available streams: vod (worst, best)
[console][info] Opening stream: vod (http)
[console][debug] Pre-buffering 8192 bytes
[console][info] Starting player: "C:\Program Files\VideoLAN\VLC\vlc.exe"
[console][debug] Writing stream to output
[console][info] Stream ended
[console][info] Closing currently open stream...```

beardypig commented:
@Billy2011 could you try changing the user agent? Eg. `--http-header "User-Agent=Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0"`

I assume you used `--player-passthrough http`? If it works in VLC, the only real difference I can think of would be the User Agent header...

Billy2011 commented:
@beardypig sorry, but it makes no difference.

back-to commented:
@Billy2011 

your *issue* can be solved with `--player-no-close`

https://streamlink.github.io/cli.html#cmdoption-player-no-close

it is not ustream related, it will always happen if you use a short video and
this is a normal behavior of streamlink
also videos will cut the last seconds if `--player-no-close` is not used

Billy2011 commented:
@beardypig sorry, but this option does not help much. After the stream ends in Streamlink, an error message appears in VLC `Can't open your input...` and then the player stalls and causes a high CPU load.

beardypig commented:
Which version of VLC are you using? Have you tried any other player?

back-to commented:
disable `repeat` in your vlc player

Billy2011 commented:
@beardypig I use only VLC and the version is 3.0.3.

@back-to, thanks for your hint, I also noticed later that `repeat` was turned on, but the playback does not start anyway as before.

Billy2011 commented:
I've installed the mpv player, but the ustream.tv VoD clips doesn't start.
I've checked it also on my linux STB, but it leads to the same result.

I've checked the playback with short YT videoclips and the clips are played; with VLC the `--player-no-close` option is needed, with MPV not.

beardypig commented:
@Billy2011, sorry for not checking properly before (I didn't have access to a real computer) - the stream appears to be malformed for streaming. You will have to download it and play it back from disk, or use the passthrough option. This is a UStream issue more than a Streamlink issue - if streamlink supported seeking for HTTP streams, then it would work :)

Billy2011 commented:
I just wanted to make it clear that it is not a problem of the player.

beardypig commented:
@liamkennedy were you able to test it out again?

liamkennedy commented:
I'm still getting inconsistent results.  

There are three variables I tested for
Using or not using 
a Different IP
--hls-live-edge 10
--http-header

First time I tried testing using the different IP (location) everything worked without even using --hls-live or --http-header or in fact using your new ustreamtv.py code at all. At that point I just could not quite believe what I was seeing. 

Then I run streamlink again and see the Reloading playlist sequence again.  So I run things a few more times just to see if the failure was now repeatable.  It failed for a few more times so I then installed your new code... then added in the --hls-edge and --http-header. When I added --hls-edge the first time it worked... then failed the second time... so I then added in the --http-header as well and it worked.  It seemed to work then for further tests with --hls-edge and --http active.   I wanted to do more testing - but had to get on to another meeting.  

Now back at my office today - things are again working inconsistently.  Right now it seems to be working reliably without --hls-edge or --http-header 

So.. to me this really does seem to point at some very weird oddness with Ustream streams.  I can't really come up with any one thing that now causes it to fail or work.

beardypig commented:
Maybe it's just an issue with their HLS server - I seem to remember way back that it used to be pretty flakey, so maybe it still sucks. 

I might try to implement the desktop version of the stream, which uses their own custom segmented mp4 system.

liamkennedy commented:
@beardypig I tend to agree on the flakiness of the HLS support being issues on the Ustream side of things.  I probably spend more time than most on monitoring the output of those streams (from NASA) and it's always has periods where it would go from flakiness to rock steady.  Just since this weekend of course it was 95% unreliable so it's been far more noticeable.  

I really appreciate you looking at implementing the desktop streams support and I know that's not an easy task.  When I first started using Livestreamer over 4 years ago they were initially supported - and the stream from NASA was the best it ever was at that time.  I know it stopped working because Ustream changed things up - and that's always the danger when dealing with an unpublished API.  

What YOU and the rest of the team here supporting Streamlink do is remarkable and I am really grateful for your efforts.

beardypig commented:
@liamkennedy thanks for kind words. 

I had a look at the desktop streams, it is fairly complicated to implement at the moment... I'm working on some changes to the code that downloads the streams, to make it more testable and to facilitate low-latency streaming. These changes should make it easier to support the custom UStream streams as well.

back-to commented:
not sure if you are looking into it,

if you use the `IPHONE` user-agent
you can get something simulary as `UHSClient` with the **websocket**
which could be more stable as the websocket maybe closes the session better than `UHSClient`

beardypig commented:
Could try using websockets... I think the HLS streams just suck, but that is something else to try - good idea @back-to

beardypig commented:
@Billy2011 I added it with a decorator (kinda) :) Normally this isn't a problem, because it's a daemon thread and Streamlink will kill it, but when using the API I can see it would be a problem :)

@liamkennedy it's that time again... I have updated the plugin to use the websockets API instead of the long polling HTTP procotol. It works better for me.

There may be some issues now with VODs or embedded streams - requires some more testing. The problem is all the old streams that I used for testing are gone now :)

Billy2011 commented:
@beardypig unfortunately urljoin () does not work with my python 2.7.9.

beardypig commented:
What do you mean it doesn’t work? That method doesn’t exist or it doesn’t like  `ws://`?

Billy2011 commented:
I think it doesn't like websocket schemes. The `urljoin()` result contains only the path without the host.

beardypig commented:
Do you think you could test it in a python shell? We only really test of the latest 2.7 version. 2.7.9 is nearly 4 years old ... are you not able to upgrade?

Billy2011 commented:
Python is part of the system image of my Linux E2 Set Top Box and I'm not the only user of streamlink. Other E2 Boxes have partly different python 2.7 versions.

beardypig commented:
Can remove it in this case, but might have to make some compatible version of urljoin - not sure if this issue also exists with other versions of python 2.7 too...

back-to commented:
only python **3.5+** will work with `ws://`

**2.7** and **3.4** won't work with it

https://github.com/python/cpython/blob/2.7/Lib/urlparse.py#L40

https://github.com/python/cpython/blob/3.4/Lib/urllib/parse.py#L43

https://github.com/python/cpython/blob/3.5/Lib/urllib/parse.py#L48

beardypig commented:
@back-to cool, I will just concatenate the strings in that case.

liamkennedy commented:
@beardypig - so you want me to try with the new ustreamtv.py here? -> https://github.com/beardypig/streamlink/blob/ustream-muxed-fix/src/streamlink/plugins/ustreamtv.py

Billy2011 commented:
@beardypig, @back-to  the python `urljoin()` issue is simple patchable:
```
ws_schemes=["ws","wss"]
(url)parse.uses_relative.extend(ws_schemes)
(url)parse.uses_netloc.extend(ws_schemes)
```

beardypig commented:
@Billy2011 best to avoid monkey patching where possible:)

beardypig commented:
@liamkennedy yep, might work better now...

Billy2011 commented:
@beardypig yeah right, but sometimes it's necessary ;)

beardypig commented:
I changed it to remove the urljoin call, does it work for you now?

liamkennedy commented:
@beardypig Latest version doesn't seem to help.  So sorry!  

```
pi@issabove-lr:~/issabove/tmp $ **streamlink http://www.ustream.tv/channel/iss-hdev-payload best --player omxplayer --fifo -l debug --hls-live-edge 10 --http-header "User-Agent=Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0"**
[console][debug] OS:         Linux-4.4.13-v7+-armv7l-with-debian-8.0
[console][debug] Python:     2.7.9
[console][debug] Streamlink: 0.13.0
[console][debug] Requests(2.4.3), Socks(1.6.7), Websocket(0.48.0)
[console][info] Found matching plugin ustreamtv for URL http://www.ustream.tv/channel/iss-hdev-payload
[plugin.ustreamtv][debug] Connecting to UStream API: media_id=17074538, application=channel, referrer=http://www.ustream.tv/channel/iss-hdev-payload, cluster=live
[plugin.ustreamtv][debug] Connecting to ws://r13940542-1-17074538-channel-ws-live.ums.ustream.tv:1935/1/ustream
[plugin.ustreamtv][debug] Sending `connect` command
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[plugin.ustreamtv][debug] Wrapping hls stream
[console][info] Available streams: 252p (worst), 360p, 486p, 720p (best)
[console][info] Opening stream: 720p (ustream)
[plugin.ustreamtv][debug] Starting API polling thread
[stream.hls][debug] Reloading playlist
[stream.hls][debug] First Sequence: 25754; Last Sequence: 25763
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 25754; End Sequence: None
[console][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 25754 to queue
[stream.hls][debug] Adding segment 25755 to queue
[stream.hls][debug] Adding segment 25756 to queue
[stream.hls][debug] Adding segment 25757 to queue
[stream.hls][debug] Adding segment 25758 to queue
[stream.hls][debug] Adding segment 25759 to queue
[stream.hls][debug] Adding segment 25760 to queue
[stream.hls][debug] Adding segment 25761 to queue
[stream.hls][debug] Adding segment 25762 to queue
[stream.hls][debug] Adding segment 25763 to queue
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007703, u'allViewers': 199083}]}
[console][info] Creating pipe streamlinkpipe-3975
[console][info] Starting player: omxplayer
[stream.hls][debug] Download of segment 25754 complete
[console][debug] Writing stream to output
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 800}]}
[stream.hls][debug] Download of segment 25755 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007705, u'allViewers': 199085}]}
[stream.hls][debug] Download of segment 25756 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007708, u'allViewers': 199088}]}
[stream.hls][debug] Download of segment 25757 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 794}]}
[stream.hls][debug] Download of segment 25758 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007710, u'allViewers': 199090}]}
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007711, u'allViewers': 199091, u'viewers': 797}]}
[stream.hls][debug] Download of segment 25759 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007715, u'allViewers': 199095}]}
[stream.hls][debug] Download of segment 25760 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007718, u'allViewers': 199098, u'viewers': 787}]}
[stream.hls][debug] Download of segment 25761 complete
[stream.hls][debug] Download of segment 25762 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007724, u'allViewers': 199104}]}
[stream.hls][debug] Download of segment 25763 complete
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 795}]}
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007727, u'allViewers': 199107}]}
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007729, u'allViewers': 199109, u'viewers': 799}]}
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 803}]}
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007731, u'allViewers': 199111}]}
[stream.hls][debug] Reloading playlist
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'viewers': 794}]}
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007737, u'allViewers': 199117}]}
[plugin.ustreamtv][debug] Received `moduleInfo` command
[plugin.ustreamtv][debug] poll response: {'cmd': u'moduleInfo', 'args': [{u'allTimeTotal': 250007740, u'allViewers': 199120}]}
[stream.hls][debug] Reloading playlist

```
keeps Reloading playlist - not further segments added/downloaded and then it quits as before

beardypig commented:
@liamkennedy blast! Could you run it with `-l trace` instead of `debug` and put the full log in a [gist](https://gist.github.com/)?

back-to commented:
@beardypig 
I think the issue is,
that **ustream sometimes** sends a wrong `#EXT-X-MEDIA-SEQUENCE:`
and streamlink can't handle it correctly

---

here is a censored version of **valid** and **invalid** playlists

https://gist.github.com/back-to/ccb9a7aa27dcf3dc7989602b5095c7e9#file-valid-m3u8

https://gist.github.com/back-to/ccb9a7aa27dcf3dc7989602b5095c7e9#file-invalid2-m3u8
https://gist.github.com/back-to/ccb9a7aa27dcf3dc7989602b5095c7e9#file-invalid-m3u8

beardypig commented:
@back-to damn, their HLS stuff really sucks :D Shame on you IBM.

back-to commented:
with this **testing** change, it works stable for me.

```diff
diff --git a/src/streamlink/stream/hls.py b/src/streamlink/stream/hls.py
index e460d46..0f7aca3 100644
--- a/src/streamlink/stream/hls.py
+++ b/src/streamlink/stream/hls.py
@@ -198,7 +198,12 @@ class HLSStreamWorker(SegmentedStreamWorker):
         if playlist.iframes_only:
             raise StreamError("Streams containing I-frames only is not playable")
 
-        media_sequence = playlist.media_sequence or 0
+        ts_num_re = re.compile(r'chunk_(?P<num>\d+)')
+        ts_num = int(ts_num_re.search(playlist.segments[0].uri).group('num'))
+        log.debug('MEDIA-SEQUENCE: {0}'.format(playlist.media_sequence))
+        log.debug('CHUNK:          {0}'.format(ts_num))
+
+        media_sequence = ts_num or 0
         sequences = [Sequence(media_sequence + i, s)
                      for i, s in enumerate(playlist.segments)]
```

---

**websocket**

**403** error for playlist after **~3 min**

I see `poll response: {'args': [{'stream': [{'name': 'uhls', 'url': 'http://` with new playlists,
but it won't update automatically / replace the old url

```
[stream.hls][debug] MEDIA-SEQUENCE: 27942
[stream.hls][debug] CHUNK:          15224
```

---

https://raw.githubusercontent.com/beardypig/streamlink/006121f0c586e254092414aeff237fd7814f8402/src/streamlink/plugins/ustreamtv.py

**runs** already for ~10 min still **no timeout**

example for the invalid numbers

```
[stream.hls][debug] MEDIA-SEQUENCE: 31216
[stream.hls][debug] CHUNK:          15105
```
and

```
[stream.hls][debug] MEDIA-SEQUENCE: 799
[stream.hls][debug] CHUNK:          15171
```

Twilight0 commented:
I am testing back-to's solution but methods like to_url() and attributes like .args are missing hense raising exceptions (TypeError, AttributeError).

back-to commented:
> (TypeError, AttributeError)

it's not set up, only `.json` works,
but the url won't work without streamlink as a ping client

beardypig commented:
I think the best solution is to implement the desktop segmented flv stream and maybe the segmented mp4 stream - but that would use ffmpeg.

Twilight0 commented:
What a tragedy, I finally got around trying the code from back-to's comment. Its nasa's ISS stream. Yes I ended up using the json object into Kodi streamlink tester, also added the custom headers directly to Kodi's ffmpeg and although the log below won't say much to you, here's what I got:

    19:27:46.727 T:3156    INFO: ffmpeg[C54]: [hls,applehttp] Opening 'http://uhs-akamai.ustream.tv/wdc04/wdc04-uhs-omega04/live/9408562/1530539788536/plain/improved/2/chunk_5166_30d9b6fede.ts' for reading
    19:27:47.102 T:3156   DEBUG: CVideoPlayer::CheckContinuity - resync forward :1, prev:17973666.666665, curr:30005666.666665, diff:12010666.666666
    19:27:47.102 T:3156   DEBUG: CVideoPlayer::CheckContinuity - resync forward :1, prev:17973666.666665, curr:30026999.999999, diff:12032000.000000
    19:27:47.102 T:3156   DEBUG: CVideoPlayer::CheckContinuity - resync forward :1, prev:17973666.666665, curr:30048333.333333, diff:12053333.333334
    19:27:47.102 T:3156   DEBUG: CVideoPlayer::CheckContinuity - resync forward :2, prev:17960333.333333, curr:30006666.666665, diff:12011666.666666
    19:27:47.102 T:3156   DEBUG: CVideoPlayer::CheckContinuity - update correction: 12011666.666666
    19:27:47.415 T:9036   DEBUG: CWinEventsWin32::WndProcWindow is active
    19:27:51.483 T:9036   DEBUG: Previous line repeats 1 times.
    19:27:51.483 T:9036   DEBUG: ------ Window Deinit (Pointer.xml) ------
    19:27:51.517 T:9036   DEBUG: ------ Window Init (Pointer.xml) ------
    19:27:59.938 T:6864  NOTICE: CVideoPlayerAudio::Process - stream stalled
    19:28:00.302 T:6876    INFO: CVideoPlayerVideo - Stillframe detected, switching to forced 30.000000 fps
    19:28:00.302 T:6876   DEBUG: CPullupCorrection: pattern lost on diff 166666.666667, number of losses 1
    19:28:04.382 T:3156   DEBUG: ffmpeg[C54]: [http] HTTP error 403 Forbidden
    19:28:04.382 T:3156   DEBUG: ffmpeg[C54]: [hls,applehttp] Failed to reload playlist 0
    19:28:04.382 T:3156   DEBUG: CVideoPlayer::SetCaching - caching state 1
    19:28:04.382 T:6876    INFO: CVideoPlayerVideo - Stillframe left, switching to normal playback
    19:28:04.382 T:6864   DEBUG: CDVDAudio::Pause - pausing audio stream
    19:28:04.382 T:3156   DEBUG: CDVDClock::SetSpeedAdjust - adjusted:0.000000
    19:28:06.179 T:9036    INFO: XCURL::DllLibCurlGlobal::CheckIdle - Closing session to http://c742ec88-lp-omega.ums.ustream.tv (easy=12B30060, multi=03F2F1B8)

After a point when Kodi is trying to reload the playlist, it gets a 403 error. LOL.

beardypig commented:
@Twilight0 as @back-to pointed out, you need streamlink to continue pinging the ustream api to maintain the hls url.

Twilight0 commented:
Perhaps I can try and ping ustream on Kodi as well.

gravyboat commented:
So what's the status on this now? Are we satisfied with the functionality or is there more you guys want to look at/do? After looking at the comment chain here it seems like we're just kind of screwed when it comes to ustream unless there are some major changes.

liamkennedy commented:
Sorry for being away from this for a couple of weeks... is there anything more I can test with the updates from @beardypig that will help improve any of the work to have Ustream (HLS) work more reliably?  Thanks for all you are doing to improve streamlink!

beardypig commented:
I think this PR fixes some bugs, but the only way to really fix it is to implement the flv/mp4 desktop streams.

back-to commented:
I think the websocket version is missing some ping back to ustream,
as they expire really fast for me.

can't really double check it,
if they need one because the mobile stream doesn't work with my webbrowser and a spoofed User-Agent

beardypig commented:
@back-to was working last time for me, but I haven’t check for a while. Maybe there is a ping pong message to respond to, but I last time I looked there wasn’t...
